spec/models/course_spec.rb

```diff
@@ -434,4 +434,103 @@
+        expect(course.enrichments.first.updated_at).to be_within(1.second).of Time.now.utc
+      end
+
+      it 'should updated last_published to the current time' do
+        course.publish_enrichment(user.id)
+        expect(course.enrichments.first.last_published_timestamp_utc).to be_within(1.second).of Time.now.utc
+      end
+
+      it 'should updated updated_by to the current user' do
+        course.publish_enrichment(user.id)
+        expect(course.enrichments.first.updated_by_user_id).to eq user.id
+      end
+    end
+
+    context 'on a course draft and published enrichments' do
+      let(:course) do
+        create(:course, with_enrichments: [
+            [:published, created_at: 5.days.ago],
+            [:published, created_at: 3.days.ago],
+            [:subsequent_draft, created_at: 1.day.ago],
+        ])
+      end
+      let(:user) { create(:user) }
+
+      it 'should publish the draft' do
+        course.publish_enrichment(user)
+        course.enrichments.each do |enrichment|
+          expect(enrichment.status).to eq 'published'
+        end
+      end
+    end
+  end
+
+  describe 'publish_sites' do
+    context 'a pre-existing course' do
+      let(:subject) { create(:course, changed_at: 10.minutes.ago) }
+      it 'updates course changed_at attribute to the current time' do
+        subject.publish_sites
+        expect(subject.changed_at).to be_within(1.second).of Time.now.utc
+      end
+    end
+
+    context 'course with a single site' do
+      let(:subject) { create(:course, with_site_statuses: [:new]) }
+
+      context 'site status is new site'do
+        it 'sets sites to running and published' do
+          subject.publish_sites
+
+          expect(subject.site_statuses.first.status).to eq 'running'
+          expect(subject.site_statuses.first.publish).to eq 'published'
+        end
+      end
+
+      it 'correctly applies all state transitions' do
+        [
+          { before: { publish: 'published', status: 'new_status' }, after: { publish: 'published', status: 'running' } },
+          { before: { publish: 'published', status: 'running' }, after: { publish: 'published', status: 'running' } },
+          { before: { publish: 'published', status: 'suspended' }, after: { publish: 'published', status: 'suspended' } },
+          { before: { publish: 'published', status: 'discontinued' }, after: { publish: 'published', status: 'discontinued' } },
+          { before: { publish: 'unpublished', status: 'new_status' }, after: { publish: 'published', status: 'running' } },
+          { before: { publish: 'unpublished', status: 'running' }, after: { publish: 'published', status: 'running' } },
+          { before: { publish: 'unpublished', status: 'suspended' }, after: { publish: 'unpublished', status: 'suspended' } },
+          { before: { publish: 'unpublished', status: 'discontinued' }, after: { publish: 'unpublished', status: 'discontinued' } },
+        ].each do |test_case|
```

Comment:
I'm not sure about this. Even though I like the idea, I think that the errors generated here will be a little hard to parse and, at the end of the day, isn't really ["unsurprising"](http://principles-wiki.net/principles:principle_of_least_surprise).

This might be clearer with a [custom matcher](https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/custom-matchers). Or, dare I say it, [shared examples](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples).

Here's a quick stab at using shared examples:

```
let(:new_published_site_status) { create :site_status, :new, :published }
let(:running_published_site_status) { create :site_status, :running, :published }

it 'changes all the site statuses' do
  expect(create :site_status, :new, :published).to be_published_and_have_status('running')
  expect(create :site_status, :running, :published).to be_published_and_have_status('running')
end
```

Reply:
rewritten, thanks for your help on this. Hope you like the replacement!